### PR TITLE
Pack localized assemblies.

### PR DESCRIPTION
**Pull request by JohanLarsson**
An attempt at #1806

### CODE REVIEW

File: src/Paket.Core/ProjectFile.fs [hunk 1]
```diff
@@ -1337,6 +1338,18 @@ type ProjectFile with
 
     member this.FindReferencesFile() = this.FindCorrespondingFile Constants.ReferencesFile
 
+    member this.FindLocalizedLanguageNames() =
+        let pattern = @"Properties\\Resources.(?<language>\w+(-\w+)?).resx"
+        let includeattribute = "Include"
+        this.ProjectNode
+        |> getDescendants "EmbeddedResource"
+        |> List.filter (fun n -> n |> hasAttribute includeattribute)
```
Comment (vasily-kirichenko) [inline src/Paket.Core/ProjectFile.fs]: why not point free?

Comment (JohanLarsson) [inline src/Paket.Core/ProjectFile.fs]: What does it mean?

Comment (theimowski) [inline src/Paket.Core/ProjectFile.fs]: means `(fun n -> n |> hasAttribute includeattribute)` is the same as `(hasAttribute includeattribute)`

Comment (JohanLarsson) [inline src/Paket.Core/ProjectFile.fs]: Thanks.

**Comment by forki** [comment]
Awesome! Only one important thing left: docs. ;-)

Could you please add a small bit about this auto-packing of satelitte assemblies?

File: src/Paket.Core/PackageMetaData.fs [hunk 2]
```diff
@@ -167,17 +167,39 @@ let findDependencies (dependenciesFile : DependenciesFile) config platform (temp
...
+                    if File.Exists fileName then
+                        let satelliteTargetDir = Path.Combine(targetDir, language)
+                        yield (FileInfo(fileName), satelliteTargetDir)
+                    else
+                        failwithf "Did not find satellite assembly for (%s) try building and running pack again." language 
+            }
+
+        let template =
+            satelliteDlls 
+            |> Seq.fold (fun template dll -> addFile (fst dll).FullName (snd dll) template) template
```
Comment (smoothdeveloper) [inline src/Paket.Core/PackageMetaData.fs]: you could decompose dll here:

```fsharp
 |> Seq.fold (fun template (dllFile, targetDir) -> addFile dllFile.FullName targetDir template) template
```

File: src/Paket.Core/PackageMetaData.fs [hunk 3]
```diff
@@ -167,17 +167,39 @@ let findDependencies (dependenciesFile : DependenciesFile) config platform (temp
...
+        let projects =                 
+            if includeReferencedProjects then 
+                project.GetAllInterProjectDependenciesWithoutProjectTemplates() 
+                |> Seq.toList 
+            else 
+                [ project ]
+
+        let satelliteDlls =
+            seq {
+            for project in projects do
```
Comment (smoothdeveloper) [inline src/Paket.Core/PackageMetaData.fs]: (minor/style) not sure of general convention, but CE block should be indented (or put `seq {` at end of previous line)

File: src/Paket.Core/PackageMetaData.fs [hunk 4]
```diff
@@ -167,17 +167,39 @@ let findDependencies (dependenciesFile : DependenciesFile) config platform (temp
...
+                let satelliteAssemblyName = Path.GetFileNameWithoutExtension(project.GetAssemblyName()) + ".resources.dll"
+                let projectDir = Path.GetDirectoryName(Path.GetFullPath(project.FileName))
+                let outputDir = Path.Combine(projectDir, project.GetOutputDirectory config platform)
+                for language in project.FindLocalizedLanguageNames() do
+                    let fileName = Path.Combine(outputDir, language, satelliteAssemblyName)
+                    if File.Exists fileName then
+                        let satelliteTargetDir = Path.Combine(targetDir, language)
+                        yield (FileInfo(fileName), satelliteTargetDir)
```
Comment (smoothdeveloper) [inline src/Paket.Core/PackageMetaData.fs]: parens are uneeded:

`yield (FileInfo filename),  satelliteTargetDir`

maybe even check if remaining are needed

File: src/Paket.Core/ProjectFile.fs [hunk 5]
```diff
@@ -1337,6 +1338,18 @@ type ProjectFile with
 
     member this.FindReferencesFile() = this.FindCorrespondingFile Constants.ReferencesFile
 
+    member this.FindLocalizedLanguageNames() =
+        let pattern = @"Properties\\Resources.(?<language>\w+(-\w+)?).resx"
+        let includeattribute = "Include"
+        this.ProjectNode
+        |> getDescendants "EmbeddedResource"
+        |> List.filter (hasAttribute includeattribute)
+        |> List.map (fun n -> n.Attributes.[includeattribute])
```
Comment (smoothdeveloper) [inline src/Paket.Core/ProjectFile.fs]: maybe define tryGetAttribute

```fsharp
let tryGetAttributeValue name node = 
    if hasAttribute name node then
        Some node.Attributes.[name].Value
    else
        None
```

then you can replace List.filter/List.map/List.map with

```fsharp
|> List.choose (tryGetAttributeValue"Include")
```

File: src/Paket.Core/ProjectFile.fs [hunk 6]
```diff
@@ -1337,6 +1338,18 @@ type ProjectFile with
...
+        |> List.map (fun a -> a.Value)
+        |> List.map (fun a -> Regex.Match(a, pattern, RegexOptions.ExplicitCapture))
```
Comment (smoothdeveloper) [inline src/Paket.Core/ProjectFile.fs]: it feels like a helper local function to handle the regex logic (return `Some "the identified language"` or `None` with single function call) would avoid mixing list operation with multistep processing of regex (which is more imperative thing).

**Comment by JohanLarsson** [comment]
@Forki updated with a mention in docs.

File: src/Paket.Core/ProjectFile.fs (hunk 6 above)
Comment (JohanLarsson) [inline src/Paket.Core/ProjectFile.fs]: Gah, I forgot about this refactoring.

File: src/Paket.Core/PackageMetaData.fs (hunk 4 above)
Comment (JohanLarsson) [inline src/Paket.Core/PackageMetaData.fs]: I could probably remove the `FileInfo` as it is just an allocation as we have already checked `Exists`

File: src/Paket.Core/PackageMetaData.fs (hunk 2 above)
Comment (JohanLarsson) [inline src/Paket.Core/PackageMetaData.fs]: Nice, I was stuck on that one for a while, used a fuzzying process.

**Comment by forki** [comment]
mhmm looks like it's broken on mono. can you please take a look?

**Comment by forki** [comment]
we also got a regression, so I reverted that change for now. see https://github.com/fsprojects/Paket/issues/1810

**Comment by JohanLarsson** [comment]
I'll be mostly adj a couple of days. Checking as soon as I can.

**Pull request merged**